File: leetcode/150/solution.go
/*
150. Evaluate Reverse Polish Notation
https://leetcode.com/problems/evaluate-reverse-polish-notation/
*/
package lc150

import "strconv"

// Implement stack
type Stack []int

func (s *Stack) Push(item int) {
	*s = append(*s, item)
}

func (s *Stack) Pop() (item int) {
	index := len(*s) - 1
	item = (*s)[index]
	*s = (*s)[:index]
	return
}

func evalRPN(tokens []string) int {
	stack := Stack{}
	// A common way to create set in golang
	set := map[string]struct{}{"+": {}, "-": {}, "*": {}, "/": {}}

	for _, t := range tokens {
		_, ok := set[t]
		if !ok {
			// t is a number -> push it to the stack
			num, _ := strconv.Atoi(t)
			stack.Push(num)
			continue
		}
		// pick up two numbers from the stack
		b := stack.Pop()
		a := stack.Pop()
		if t == "+" {
			stack.Push(a + b)
		} else if t == "-" {
			stack.Push(a - b)
		} else if t == "*" {
			stack.Push(a * b)
		} else {
			sign := 1
			if a < 0 {
				a, sign = a*-1, -1
			}
			if b < 0 {
				b, sign = b*-1, sign*-1
			}
			stack.Push(a / b * sign)
		}
	}

	return stack[0]
}

File: leetcode/150/solution.py
"""
150. Evaluate Reverse Polish Notation
https://leetcode.com/problems/evaluate-reverse-polish-notation/
"""
from typing import List, Tuple
from unittest import TestCase, main


class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack: List[int] = []
        operator = ("+", "-", "*", "/")

        for t in tokens:
            if t not in operator:
                # t is a number -> push it to the stack
                stack.append(int(t))
                continue
            # pick up two numbers from  the stack
            b = stack.pop()
            a = stack.pop()
            if t == "+":
                stack.append(a + b)
            elif t == "-":
                stack.append(a - b)
            elif t == "*":
                stack.append(a * b)
            else:
                sign = 1
                if a < 0:
                    a, sign = -a, -1
                if b < 0:
                    b, sign = -b, sign * -1
                stack.append(a // b * sign)

        return stack[0]


class Test(TestCase):
    data: List[Tuple[List[str], int]] = [
        # (["2", "1", "+", "3", "*"], 9),
        # (["4", "13", "5", "/", "+"], 6),
        (["10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"], 22),
    ]

    def test_solution(self):
        s = Solution()
        for input, expected in self.data:
            self.assertEqual(s.evalRPN(input), expected)


if __name__ == "__main__":
    main()
